### PR TITLE
ci: release pipeline

### PR DESCRIPTION
**Pull request by kuisathaverat**
Pipeline to rollout a deployment with the latest version of the Package registry. The user selects the deployment name (snapshot, staging, or production), then this deployment is rollout with the latest version.

### CODE REVIEW

File: .ci/release.groovy [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    disableConcurrentBuilds()
+  }
+  parameters {
+    choice(choices: ['snapshot', 'staging', 'production'], description: 'Branch and environment to Release the distribution.', name: 'environment')
+    string(name: 'commit', defaultValue: "", description: "Commit SHA1/Branch tag of the Docker image.")
+  }
+  stages {
+    stage('Select Docker Image'){
+      when {
+        expression { return '' == params.commit }
```
Comment (v1v) [inline .ci/release.groovy:50]: I was getting a bit lost with this, what do you think to use the env instead?
```suggestion
        expression { return '' == env.TAG }
```

**Comment by ruflin** [comment]
I probably forgot something during my days off, but why don't we make staging and snapshot just rely on on `distribution:snapshot` and trigger a new rollout?

File: .ci/release.groovy [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+              yml.image.tag = "${env.TAG}"
+              writeYaml(file: ymlFile, data: yml, overwrite: true)
+              setEnvVar('DOCKER_IMAGE', yml.image.repository)
+            }
+            sh(label: 'Git config', script: """
+              git config --global user.name observability-robots
+              git config --global user.email observability-robots@users.noreply.github.com
+            """)
+            sh(label: 'Commit local changes', script: "git commit -a -m '[package-registry]: Release ${env.TAG} to ${env.BRANCH_NAME}'")
+            sh(label: 'Push changes', script: "git push origin package-registry-${env.REPO}-release --force")
```
Comment (v1v) [inline .ci/release.groovy:102]: Could it be replaced with the below steps?
- `withGitRelease()` 
-  `gitPush()`

File: .ci/release.groovy [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            sh(label: 'Check Docker image', script: "docker pull ${env.DOCKER_IMAGE}:${env.TAG}")
+            githubCreatePullRequest(
+              title: "[DO NO MERGE][package-registry]: Release ${env.TAG} to ${env.BRANCH_NAME}",
+              base: "elastic:master",
+              labels: 'area:web',
+              assign: 'elasticmachine',
+              reviewer: 'observability-robots',
+              description: """
+This is an automatic generated PR to release `docker.elastic.co/package-registry/distribution:${env.TAG}` to ${env.BRANCH_NAME}.
+  * [ ] Check tge TAG ${env.TAG} and the enviroment ${env.BRANCH_NAME} is correct.
```
Comment (cachedout) [inline .ci/release.groovy:112]:
```suggestion
  * [ ] Check the TAG ${env.TAG} and the enviroment ${env.BRANCH_NAME} is correct.
```

**Comment by kuisathaverat** [comment]
>I probably forgot something during my days off, but why don't we make staging and snapshot just rely on on distribution:snapshot and trigger a new rollout?

@ruflin 
The rollout is made by the automation made in place by the infra team to redeploy changes on the Helm chart, the idea is to reuse this logic and make the same kind deploy for snapshot, staging, and production. As I commented this is the first iteration, and we have to confirm with infra team that we can put something in place that auto-merge the PRs for the snapshot and staging environments, in that way, the deploy on snapshot and staging is a single button deploy that it is the goal.

**Comment by ruflin** [comment]
@kuisathaverat For now I think just having 1 should be enough.

**Comment by mtojek** [comment]
@ruflin this is relatively old PR. Can we close this one or is there a bigger initiative behind?

**Comment by ruflin** [comment]
Same as the other CI PR, @kuisathaverat can you comment on it?

**Comment by kuisathaverat** [comment]
@ruflin I've changed the approach to make a rollout. This job is triggered manually when it works we would add a new stage on the current pipelines to trigger it on merges to branches, with approval if it is needed

File: .ci/release.groovy [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            ''')
+          }
+        }
+      }
+    }
+  }
+}
+
+def getVaultSecretRetry(Map args){
+  def secret = arg.hasKey('') ? args.secret : error('Secret not valid')
```
Comment (v1v) [inline .ci/release.groovy:70]:
```suggestion
  def secret = arg.containsKey('secret') ? args.secret : error('Secret not valid')
```

File: .ci/release.groovy [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+def withPackageRegistryEnv(Map args, Closure body){
+  def jsonValue = getVaultSecretRetry(args)
+  withEnvMask(vars: [
+    [var: "GOOGLE_PROJECT", password: jsonValue.google_project],
+    [var: "REGION", password: jsonValue.region],
+    [var: "CLUSTER_CREDENTIALS_NAME", password: jsonValue.cluster_credentials_name],
+    [var: "KUBECONFIG", password: "${HOME}/.kubeconfig"],
+  ]){
+    body()
```
Comment (v1v) [inline .ci/release.groovy:86]: I'd rather prefer to configure the installation and env variable in this 
```suggestion
    def k8sLocation = "$HOME/bin"
    sh(label: 'Install Kubectl', script: '''
      curl -Lo $HOME/bin/kubectl https://storage.googleapis.com/kubernetes-release/release/v1.19.0/bin/linux/amd64/kubectl
      chmod +x ${k8sLocation}/kubectl''')
    withEnv(["PATH+K8S=${k8sLocation}"]) {
      body()
    }
```

File: .ci/release.groovy [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  ''')
+  body()
+  sh(label: 'delete credentials', script: 'rm -fr ${CREDENTIALS_FILE} ${KUBECONFIG}')
+}
+
+def installKubectl(){
+  sh(label: 'Install Kubectl', script: '''
+  curl -Lo $HOME/bin/kubectl https://storage.googleapis.com/kubernetes-release/release/v1.19.0/bin/linux/amd64/kubectl
+  chmod +x $HOME/bin/kubectl
+  ''')
```
Comment (v1v) [inline .ci/release.groovy:105]:
```suggestion
```

File: .ci/release.groovy [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+pipeline {
+  agent { label 'linux && immutable' }
+  environment {
+    REPO = "package-storage"
+    NOTIFY_TO = credentials('notify-to')
+    PIPELINE_LOG_LEVEL = 'INFO'
+    LANG = "C.UTF-8"
+    LC_ALL = "C.UTF-8"
+    HOME = "${env.WORKSPACE}"
+    PATH = "${env.HOME}/bin:${env.PATH}"
```
Comment (v1v) [inline .ci/release.groovy:29]:
```suggestion
```

File: .ci/release.groovy [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+          return params.environment != 'none'
+        }
+      }
+      environment{
+        PACKAGE_REGISTRY_DEPLOYMENT_NAME = "package-registry-${params.environment}-vanilla"
+      }
+      steps {
+        withPackageRegistryEnv(secret: 'secret/observability-team/ci/package-registry-deployment'){
+          withGCPCredentials(secret: "secret/gce/${GOOGLE_PROJECT}/service-account/package-registry-rollout"){
+            installKubectl()
```
Comment (v1v) [inline .ci/release.groovy:58]:
```suggestion
```

**Review by v1v — COMMENTED** [review]
NIT, regarding the PATH and a missing validation

File: .ci/release.groovy (hunk 7 above)
Comment (kuisathaverat) [inline .ci/release.groovy:29]:
```suggestion
    PATH = "${env.HOME}/bin:${env.PATH}"
    KUBECTL = "${env.HOME}/bin/kubectl"
```

File: .ci/release.groovy [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            sh(label: "Rollout ${PACKAGE_REGISTRY_DEPLOYMENT_NAME} deployment", script: '''
+              kubectl -n package-registry rollout restart deployment ${PACKAGE_REGISTRY_DEPLOYMENT_NAME}
```
Comment (kuisathaverat) [inline .ci/release.groovy:60]:
```suggestion
              ${KUBECTL} -n package-registry rollout restart deployment ${PACKAGE_REGISTRY_DEPLOYMENT_NAME}
```

File: .ci/release.groovy [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    gcloud auth activate-service-account --key-file ${CREDENTIALS_FILE}
+    gcloud --project=${GOOGLE_PROJECT} container clusters get-credentials ${CLUSTER_CREDENTIALS_NAME} --region ${REGION}
+  ''')
+  body()
+  sh(label: 'delete credentials', script: 'rm -fr ${CREDENTIALS_FILE} ${KUBECONFIG}')
+}
+
+def installKubectl(){
+  sh(label: 'Install Kubectl', script: '''
+  curl -Lo $HOME/bin/kubectl https://storage.googleapis.com/kubernetes-release/release/v1.19.0/bin/linux/amd64/kubectl
```
Comment (kuisathaverat) [inline .ci/release.groovy:103]:
```suggestion
  curl -Lo ${KUBECTL} https://storage.googleapis.com/kubernetes-release/release/v1.19.0/bin/linux/amd64/kubectl
```

File: .ci/release.groovy [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+  chmod +x $HOME/bin/kubectl
```
Comment (kuisathaverat) [inline .ci/release.groovy:104]:
```suggestion
  chmod +x ${KUBECTL}
```

File: .ci/release.groovy (hunk 5 above)
Comment (kuisathaverat) [reply .ci/release.groovy:86]: a withEnv inside a withEnvMask 🤔 maybe put Path definition on the withEnvMask, but it is not needed anymore

**Comment by ruflin** [comment]
@kuisathaverat So after this is merged, the rollout can be done by going to the Jenkins UI?

**Comment by kuisathaverat** [comment]
> @kuisathaverat So after this is merged, the rollout can be done by going to the Jenkins UI?

yes, you select the environment and push build 😄

File: .ci/release.groovy [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+      }
+    }
+  }
+}
+
+def getVaultSecretRetry(Map args){
+  def secret = arg.containsKey('secret') ? args.secret : error('Secret not valid')
+  def jsonValue = [:]
+  retryWithSleep(retries: 3, seconds: 5, backoff: true) {
+    jsonValue = getVaultSecret(secret: secret)
```
Comment (cachedout) [inline .ci/release.groovy:73]: Isn't there [already a retry in this function](https://github.com/elastic/apm-pipeline-library/blob/master/vars/getVaultSecret.groovy#L51)?

Comment (kuisathaverat) [reply .ci/release.groovy:73]: 🤦 yep, I can remove the retry

File: .ci/release.groovy [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+  }
```
Comment (kuisathaverat) [inline .ci/release.groovy:74]:
```suggestion
  jsonValue = getVaultSecret(secret: secret)
```

File: .ci/release.groovy [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+      }
+    }
+  }
+}
+
+def getVaultSecretRetry(Map args){
+  def secret = arg.containsKey('secret') ? args.secret : error('Secret not valid')
+  def jsonValue = [:]
+  jsonValue = getVaultSecret(secret: secret)
+  return jsonValue
```
Comment (kuisathaverat) [inline .ci/release.groovy:73]:
```suggestion
  return getVaultSecret(secret: secret)
```

File: .ci/release.groovy [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+          }
+        }
+      }
+    }
+  }
+}
+
+def getVaultSecretRetry(Map args){
+  def secret = arg.containsKey('secret') ? args.secret : error('Secret not valid')
+  def jsonValue = [:]
```
Comment (kuisathaverat) [inline .ci/release.groovy:71]:
```suggestion
```

**Pull request merged**